{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "national-custody",
   "metadata": {
    "heading_collapsed": "true"
   },
   "source": [
    "# Introduction"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "running-steal",
   "metadata": {},
   "source": [
    "This is yet another tool for annotating genomic regions. Why was that necessary?\n",
    "\n",
    "1. This tool provides more detailed information than we found in existing tools.\n",
    "2. We leverage all this information to resolve multiple potential annotations for the same regions into primary annotations using a set of precedence rules. This gives a single annotation per region, with a clear decision tree for each choice.\n",
    "3. This tool provides a more flexibel definition of regions of interest with regard to TSS than we found in existing tools. For example, we like to partition the regions surrounding the TSS into a custom promoter region (say 5000 upstream and 1000 bp downstream of the TSS), and an adjacent more distant cis-regulatory region upstream of the promoter (maybe up to 45000 bp upstream of the promoter region). That's easy with gtfanno.\n",
    "\n",
    "The tool provides the following basic annotations in this order of precedence (by default):\n",
    "- Promoter\n",
    "- 5'UTR, 3'UTR\n",
    "- Intron, Exon\n",
    "- Distant cis-regulatory region (see above)\n",
    "- Intergenic\n",
    "\n",
    "The basic annotation workflow is as follows\n",
    "- For each region, find all possible annotations. An annotation is considered if the center of the region lies within the feature.\n",
    "- For all annotations, various statistics are recorded, such as the number and percentages of overlapping bp or the distance to the TSS of the gene.\n",
    "- For each region, define the primary feature class (e.g. Promoter, or Intron).\n",
    "- Optionally, select the preferred annotation within the primary feature class according to various statistics. For example, a region may be annotated to the promoters of multiple genes. The top annotation could be chosen by the distance of the region center to the TSS, or by the percentage of overlapping bp.\n",
    "\n",
    "This workflow produces the following output files:\n",
    "\n",
    "Output files:\n",
    "  - all_annotations (as BED3+ and pickled DataFrame)\n",
    "    - contains all annotations\n",
    "  - primary_annotations (BED3+ and pickled DataFrame):\n",
    "    - contains only primary annotations for each region\n",
    "\n",
    "Optional selection of the top annotation for each gene can be performed with simple dataframe operations based on the primary_annotations dataframe. See below for an example!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "textile-reserve",
   "metadata": {
    "heading_collapsed": "true"
   },
   "source": [
    "# Package maturity"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "monthly-pierce",
   "metadata": {},
   "source": [
    "This package is unreleased and unpublished software, but we use it often in in-house projects. We can not yet provide support for external users. Also, given the development status, we do change the API from time to time, without regard for health and safety of external users."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "engaging-realtor",
   "metadata": {},
   "source": [
    "# Usage"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "satellite-heather",
   "metadata": {},
   "source": [
    "## Import"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "thousand-opportunity",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "import gtfanno\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "subsequent-madison",
   "metadata": {},
   "source": [
    "## Input data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "advisory-grass",
   "metadata": {},
   "source": [
    "The genomic regions to be annotated are expected in BED3 format. A header line is allowed if commented with '#'. Additional columns are ignored. The file must be sorted `-k1,1 -k2,2n -k3,3n`. The code is currently only tested with chromosome names without prefix (i.e. '1', not 'chr1')."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "brave-sharing",
   "metadata": {},
   "source": [
    "Here are some random demo regions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "missing-kingston",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#chrom\tstart\tend\n",
      "1\t7139742\t7139990\n",
      "1\t9548071\t9548115\n",
      "1\t10316090\t10316863\n",
      "1\t16297792\t16297835\n",
      "1\t33482222\t33482405\n",
      "1\t33594864\t33596335\n",
      "1\t34007363\t34007556\n",
      "1\t34029970\t34030115\n",
      "1\t34130946\t34131079\n"
     ]
    }
   ],
   "source": [
    "!head demo-regions.bed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "adjacent-native",
   "metadata": {},
   "source": [
    "## Performing the gene annotation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "valuable-norfolk",
   "metadata": {},
   "source": [
    "Most users will only need a single command to use this package. This command is not yet optimized and builds some internal databases at each execution - plan a few minutes of time for this."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "infinite-zimbabwe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading data\n",
      "Annotating promoter regions\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/stephen/miniconda3/envs/gtfanno_env/lib/python3.8/subprocess.py:853: RuntimeWarning: line buffering (buffering=1) isn't supported in binary mode, the default buffer size will be used\n",
      "  self.stderr = io.open(errread, 'rb', bufsize)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Annotating transcript parts\n",
      "UTR classification\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/stephen/miniconda3/envs/gtfanno_env/lib/python3.8/subprocess.py:853: RuntimeWarning: line buffering (buffering=1) isn't supported in binary mode, the default buffer size will be used\n",
      "  self.stderr = io.open(errread, 'rb', bufsize)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Merge results\n",
      "Classify annotations\n",
      "Add intergenic regions\n",
      "Save results\n",
      "Basic stats for primary annotations\n",
      "   #Primary annotations  Frequency\n",
      "0                     1       2497\n",
      "1                     2        336\n",
      "2                     3        112\n",
      "3                     4         40\n",
      "4                     5          7\n",
      "5                     6          5\n",
      "6                    22          1\n",
      "7                     7          1\n",
      "8                     9          1\n",
      "      Feature  Frequency\n",
      "0      intron       1485\n",
      "1        DCRD       1284\n",
      "2    Promoter        616\n",
      "3  intergenic        265\n",
      "4        exon         65\n",
      "5      3'-UTR         49\n",
      "6      5'-UTR          4\n",
      "7  transcript          0\n",
      "CPU times: user 4min 32s, sys: 884 ms, total: 4min 33s\n",
      "Wall time: 4min 34s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "os.makedirs(\"results\", exist_ok=True)\n",
    "gtfanno.annotate(\n",
    "    query_bed=\"demo-regions.bed\",\n",
    "    # Path to a gene annotation in GTF format\n",
    "    gtf_fp=\"gencode.vM26.basic.annotation.no-chr-prefix.gtf.gz\",\n",
    "    # all results are placed in results/demo-regions_{suffix}\n",
    "    trunk_path=\"results/demo-regions\",\n",
    "    tmpdir=\"/tmp\",\n",
    "    # upstream, downstream from TSS\n",
    "    promoter=(-5000, 1000),\n",
    "    # annotation of a second, broader window around the TSS is often helpful\n",
    "    distant_cis_regulatory_domain=(-50_000, -5000),\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "stock-darwin",
   "metadata": {},
   "source": [
    "## Output data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "functioning-bruce",
   "metadata": {},
   "source": [
    "This command produces the output files for all annotations and primary annotations described in the [introduction](#Introduction). Both files are available as pickle and BED files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "talented-insulation",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 4.9M\n",
      "-rw-rw-r-- 1 stephen stephen 1.9M Feb 21 17:34 demo-regions_all-annotations.bed\n",
      "-rw-rw-r-- 1 stephen stephen 1.8M Feb 21 17:34 demo-regions_all-annotations.p\n",
      "-rw-rw-r-- 1 stephen stephen 600K Feb 21 17:34 demo-regions_primary-annotations.bed\n",
      "-rw-rw-r-- 1 stephen stephen 619K Feb 21 17:34 demo-regions_primary-annotations.p\n"
     ]
    }
   ],
   "source": [
    "!ls -lh results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "threatened-floor",
   "metadata": {},
   "source": [
    "The resulting dataframes have various annotations, for example a 'distance' annotation. This gives the distance `feature center - TSS` for promoter, DCRD and intergenic annotations, and the distance `center of the feature - center of the region` otherwise."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "behind-bronze",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>gtfanno_uid</th>\n",
       "      <th>center</th>\n",
       "      <th>feat_class</th>\n",
       "      <th>perc_feature</th>\n",
       "      <th>perc_region</th>\n",
       "      <th>distance</th>\n",
       "      <th>has_center</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>appris_principal_score</th>\n",
       "      <th>feat_chrom</th>\n",
       "      <th>feat_start</th>\n",
       "      <th>feat_end</th>\n",
       "      <th>feat_center</th>\n",
       "      <th>feat_strand</th>\n",
       "      <th>feature_rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>7139742</td>\n",
       "      <td>7139990</td>\n",
       "      <td>0</td>\n",
       "      <td>7139866.0</td>\n",
       "      <td>DCRD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-19278.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Pcmtd1</td>\n",
       "      <td>ENSMUSG00000051285.18</td>\n",
       "      <td>ENSMUST00000061280.17</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7109144.0</td>\n",
       "      <td>7160144.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>9548071</td>\n",
       "      <td>9548115</td>\n",
       "      <td>1</td>\n",
       "      <td>9548093.0</td>\n",
       "      <td>DCRD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-37492.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Gm18299</td>\n",
       "      <td>ENSMUSG00000104428.2</td>\n",
       "      <td>ENSMUST00000192026.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9509601.0</td>\n",
       "      <td>9560601.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>10316090</td>\n",
       "      <td>10316863</td>\n",
       "      <td>2</td>\n",
       "      <td>10316476.5</td>\n",
       "      <td>DCRD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-47449.5</td>\n",
       "      <td>True</td>\n",
       "      <td>Gm37569</td>\n",
       "      <td>ENSMUSG00000102556.2</td>\n",
       "      <td>ENSMUST00000192111.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10268027.0</td>\n",
       "      <td>10319027.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>10316090</td>\n",
       "      <td>10316863</td>\n",
       "      <td>2</td>\n",
       "      <td>10316476.5</td>\n",
       "      <td>DCRD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-13581.5</td>\n",
       "      <td>True</td>\n",
       "      <td>Arfgef1</td>\n",
       "      <td>ENSMUSG00000067851.12</td>\n",
       "      <td>ENSMUST00000088615.11</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10301895.0</td>\n",
       "      <td>10352895.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>10316090</td>\n",
       "      <td>10316863</td>\n",
       "      <td>2</td>\n",
       "      <td>10316476.5</td>\n",
       "      <td>DCRD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-13581.5</td>\n",
       "      <td>True</td>\n",
       "      <td>Arfgef1</td>\n",
       "      <td>ENSMUSG00000067851.12</td>\n",
       "      <td>ENSMUST00000131556.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10301895.0</td>\n",
       "      <td>10352895.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>secondary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>16297792</td>\n",
       "      <td>16297835</td>\n",
       "      <td>3</td>\n",
       "      <td>16297813.5</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-916.5</td>\n",
       "      <td>True</td>\n",
       "      <td>Gm28095</td>\n",
       "      <td>ENSMUSG00000100868.2</td>\n",
       "      <td>ENSMUST00000186343.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>16248730.0</td>\n",
       "      <td>16299730.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>16297792</td>\n",
       "      <td>16297835</td>\n",
       "      <td>3</td>\n",
       "      <td>16297813.5</td>\n",
       "      <td>DCRD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-26002.5</td>\n",
       "      <td>True</td>\n",
       "      <td>Gm7568</td>\n",
       "      <td>ENSMUSG00000089982.8</td>\n",
       "      <td>ENSMUST00000161626.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>16273816.0</td>\n",
       "      <td>16324816.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>secondary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>16297792</td>\n",
       "      <td>16297835</td>\n",
       "      <td>3</td>\n",
       "      <td>16297813.5</td>\n",
       "      <td>DCRD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-26002.5</td>\n",
       "      <td>True</td>\n",
       "      <td>Gm7568</td>\n",
       "      <td>ENSMUSG00000089982.8</td>\n",
       "      <td>ENSMUST00000160446.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>16273816.0</td>\n",
       "      <td>16324816.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>secondary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>16297792</td>\n",
       "      <td>16297835</td>\n",
       "      <td>3</td>\n",
       "      <td>16297813.5</td>\n",
       "      <td>DCRD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-17072.5</td>\n",
       "      <td>True</td>\n",
       "      <td>Gm38249</td>\n",
       "      <td>ENSMUSG00000104165.2</td>\n",
       "      <td>ENSMUST00000192560.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>16279741.0</td>\n",
       "      <td>16330741.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>secondary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>33482222</td>\n",
       "      <td>33482405</td>\n",
       "      <td>4</td>\n",
       "      <td>33482313.5</td>\n",
       "      <td>DCRD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-38868.5</td>\n",
       "      <td>True</td>\n",
       "      <td>Gm29228</td>\n",
       "      <td>ENSMUSG00000101682.2</td>\n",
       "      <td>ENSMUST00000185489.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>33471182.0</td>\n",
       "      <td>33522182.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1</td>\n",
       "      <td>33594864</td>\n",
       "      <td>33596335</td>\n",
       "      <td>5</td>\n",
       "      <td>33595599.0</td>\n",
       "      <td>transcript</td>\n",
       "      <td>0.006811</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-5284.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Prim2</td>\n",
       "      <td>ENSMUSG00000026134.12</td>\n",
       "      <td>ENSMUST00000027312.11</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>33492891.0</td>\n",
       "      <td>33708876.0</td>\n",
       "      <td>33600883.0</td>\n",
       "      <td>-</td>\n",
       "      <td>primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1</td>\n",
       "      <td>34007363</td>\n",
       "      <td>34007556</td>\n",
       "      <td>6</td>\n",
       "      <td>34007459.5</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-4693.5</td>\n",
       "      <td>True</td>\n",
       "      <td>Gm28631</td>\n",
       "      <td>ENSMUSG00000099382.2</td>\n",
       "      <td>ENSMUST00000190740.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>34001766.0</td>\n",
       "      <td>34052766.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1</td>\n",
       "      <td>34007363</td>\n",
       "      <td>34007556</td>\n",
       "      <td>6</td>\n",
       "      <td>34007459.0</td>\n",
       "      <td>transcript</td>\n",
       "      <td>0.000483</td>\n",
       "      <td>1.0</td>\n",
       "      <td>139530.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Dst</td>\n",
       "      <td>ENSMUSG00000026131.21</td>\n",
       "      <td>ENSMUST00000183034.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>33947306.0</td>\n",
       "      <td>34346673.0</td>\n",
       "      <td>34146989.0</td>\n",
       "      <td>+</td>\n",
       "      <td>secondary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1</td>\n",
       "      <td>34007363</td>\n",
       "      <td>34007556</td>\n",
       "      <td>6</td>\n",
       "      <td>34007459.5</td>\n",
       "      <td>DCRD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-43446.5</td>\n",
       "      <td>True</td>\n",
       "      <td>Dst</td>\n",
       "      <td>ENSMUSG00000026131.21</td>\n",
       "      <td>ENSMUST00000115104.10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>34000906.0</td>\n",
       "      <td>34051906.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>secondary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1</td>\n",
       "      <td>34007363</td>\n",
       "      <td>34007556</td>\n",
       "      <td>6</td>\n",
       "      <td>34007459.5</td>\n",
       "      <td>DCRD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-43446.5</td>\n",
       "      <td>True</td>\n",
       "      <td>Dst</td>\n",
       "      <td>ENSMUSG00000026131.21</td>\n",
       "      <td>ENSMUST00000097785.10</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>34000906.0</td>\n",
       "      <td>34051906.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>secondary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1</td>\n",
       "      <td>34007363</td>\n",
       "      <td>34007556</td>\n",
       "      <td>6</td>\n",
       "      <td>34007459.5</td>\n",
       "      <td>DCRD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-43434.5</td>\n",
       "      <td>True</td>\n",
       "      <td>Dst</td>\n",
       "      <td>ENSMUSG00000026131.21</td>\n",
       "      <td>ENSMUST00000097786.10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>34000894.0</td>\n",
       "      <td>34051894.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>secondary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1</td>\n",
       "      <td>34007363</td>\n",
       "      <td>34007556</td>\n",
       "      <td>6</td>\n",
       "      <td>34007459.5</td>\n",
       "      <td>DCRD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-37747.5</td>\n",
       "      <td>True</td>\n",
       "      <td>Dst</td>\n",
       "      <td>ENSMUSG00000026131.21</td>\n",
       "      <td>ENSMUST00000182314.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>33995207.0</td>\n",
       "      <td>34046207.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>secondary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1</td>\n",
       "      <td>34007363</td>\n",
       "      <td>34007556</td>\n",
       "      <td>6</td>\n",
       "      <td>34007459.5</td>\n",
       "      <td>DCRD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-37503.5</td>\n",
       "      <td>True</td>\n",
       "      <td>Dst</td>\n",
       "      <td>ENSMUSG00000026131.21</td>\n",
       "      <td>ENSMUST00000182697.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>33994963.0</td>\n",
       "      <td>34045963.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>secondary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1</td>\n",
       "      <td>34029970</td>\n",
       "      <td>34030115</td>\n",
       "      <td>7</td>\n",
       "      <td>34030042.0</td>\n",
       "      <td>transcript</td>\n",
       "      <td>0.000363</td>\n",
       "      <td>1.0</td>\n",
       "      <td>116947.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Dst</td>\n",
       "      <td>ENSMUSG00000026131.21</td>\n",
       "      <td>ENSMUST00000183034.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>33947306.0</td>\n",
       "      <td>34346673.0</td>\n",
       "      <td>34146989.0</td>\n",
       "      <td>+</td>\n",
       "      <td>primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1</td>\n",
       "      <td>34029970</td>\n",
       "      <td>34030115</td>\n",
       "      <td>7</td>\n",
       "      <td>34030042.5</td>\n",
       "      <td>DCRD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-41138.5</td>\n",
       "      <td>True</td>\n",
       "      <td>Gm38157</td>\n",
       "      <td>ENSMUSG00000103364.2</td>\n",
       "      <td>ENSMUST00000194178.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>34021181.0</td>\n",
       "      <td>34072181.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>secondary</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome     Start       End  gtfanno_uid      center  feat_class  \\\n",
       "0           1   7139742   7139990            0   7139866.0        DCRD   \n",
       "1           1   9548071   9548115            1   9548093.0        DCRD   \n",
       "2           1  10316090  10316863            2  10316476.5        DCRD   \n",
       "3           1  10316090  10316863            2  10316476.5        DCRD   \n",
       "4           1  10316090  10316863            2  10316476.5        DCRD   \n",
       "5           1  16297792  16297835            3  16297813.5    Promoter   \n",
       "6           1  16297792  16297835            3  16297813.5        DCRD   \n",
       "7           1  16297792  16297835            3  16297813.5        DCRD   \n",
       "8           1  16297792  16297835            3  16297813.5        DCRD   \n",
       "9           1  33482222  33482405            4  33482313.5        DCRD   \n",
       "10          1  33594864  33596335            5  33595599.0  transcript   \n",
       "11          1  34007363  34007556            6  34007459.5    Promoter   \n",
       "12          1  34007363  34007556            6  34007459.0  transcript   \n",
       "13          1  34007363  34007556            6  34007459.5        DCRD   \n",
       "14          1  34007363  34007556            6  34007459.5        DCRD   \n",
       "15          1  34007363  34007556            6  34007459.5        DCRD   \n",
       "16          1  34007363  34007556            6  34007459.5        DCRD   \n",
       "17          1  34007363  34007556            6  34007459.5        DCRD   \n",
       "18          1  34029970  34030115            7  34030042.0  transcript   \n",
       "19          1  34029970  34030115            7  34030042.5        DCRD   \n",
       "\n",
       "    perc_feature  perc_region  distance has_center gene_name  \\\n",
       "0            NaN          NaN  -19278.0       True    Pcmtd1   \n",
       "1            NaN          NaN  -37492.0       True   Gm18299   \n",
       "2            NaN          NaN  -47449.5       True   Gm37569   \n",
       "3            NaN          NaN  -13581.5       True   Arfgef1   \n",
       "4            NaN          NaN  -13581.5       True   Arfgef1   \n",
       "5            NaN          NaN    -916.5       True   Gm28095   \n",
       "6            NaN          NaN  -26002.5       True    Gm7568   \n",
       "7            NaN          NaN  -26002.5       True    Gm7568   \n",
       "8            NaN          NaN  -17072.5       True   Gm38249   \n",
       "9            NaN          NaN  -38868.5       True   Gm29228   \n",
       "10      0.006811          1.0   -5284.0       True     Prim2   \n",
       "11           NaN          NaN   -4693.5       True   Gm28631   \n",
       "12      0.000483          1.0  139530.0       True       Dst   \n",
       "13           NaN          NaN  -43446.5       True       Dst   \n",
       "14           NaN          NaN  -43446.5       True       Dst   \n",
       "15           NaN          NaN  -43434.5       True       Dst   \n",
       "16           NaN          NaN  -37747.5       True       Dst   \n",
       "17           NaN          NaN  -37503.5       True       Dst   \n",
       "18      0.000363          1.0  116947.0       True       Dst   \n",
       "19           NaN          NaN  -41138.5       True   Gm38157   \n",
       "\n",
       "                  gene_id          transcript_id  appris_principal_score  \\\n",
       "0   ENSMUSG00000051285.18  ENSMUST00000061280.17                     1.0   \n",
       "1    ENSMUSG00000104428.2   ENSMUST00000192026.2                     0.0   \n",
       "2    ENSMUSG00000102556.2   ENSMUST00000192111.2                     0.0   \n",
       "3   ENSMUSG00000067851.12  ENSMUST00000088615.11                     1.0   \n",
       "4   ENSMUSG00000067851.12   ENSMUST00000131556.2                     0.0   \n",
       "5    ENSMUSG00000100868.2   ENSMUST00000186343.2                     0.0   \n",
       "6    ENSMUSG00000089982.8   ENSMUST00000161626.2                     0.0   \n",
       "7    ENSMUSG00000089982.8   ENSMUST00000160446.8                     0.0   \n",
       "8    ENSMUSG00000104165.2   ENSMUST00000192560.2                     0.0   \n",
       "9    ENSMUSG00000101682.2   ENSMUST00000185489.2                     0.0   \n",
       "10  ENSMUSG00000026134.12  ENSMUST00000027312.11                     1.0   \n",
       "11   ENSMUSG00000099382.2   ENSMUST00000190740.2                     0.0   \n",
       "12  ENSMUSG00000026131.21   ENSMUST00000183034.5                     0.0   \n",
       "13  ENSMUSG00000026131.21  ENSMUST00000115104.10                     0.0   \n",
       "14  ENSMUSG00000026131.21  ENSMUST00000097785.10                     3.0   \n",
       "15  ENSMUSG00000026131.21  ENSMUST00000097786.10                     0.0   \n",
       "16  ENSMUSG00000026131.21   ENSMUST00000182314.8                     0.0   \n",
       "17  ENSMUSG00000026131.21   ENSMUST00000182697.8                     0.0   \n",
       "18  ENSMUSG00000026131.21   ENSMUST00000183034.5                     0.0   \n",
       "19   ENSMUSG00000103364.2   ENSMUST00000194178.2                     0.0   \n",
       "\n",
       "    feat_chrom  feat_start    feat_end  feat_center feat_strand feature_rank  \n",
       "0          1.0   7109144.0   7160144.0          NaN           +      primary  \n",
       "1          1.0   9509601.0   9560601.0          NaN           -      primary  \n",
       "2          1.0  10268027.0  10319027.0          NaN           -      primary  \n",
       "3          1.0  10301895.0  10352895.0          NaN           -      primary  \n",
       "4          1.0  10301895.0  10352895.0          NaN           -    secondary  \n",
       "5          1.0  16248730.0  16299730.0          NaN           +      primary  \n",
       "6          1.0  16273816.0  16324816.0          NaN           +    secondary  \n",
       "7          1.0  16273816.0  16324816.0          NaN           +    secondary  \n",
       "8          1.0  16279741.0  16330741.0          NaN           -    secondary  \n",
       "9          1.0  33471182.0  33522182.0          NaN           +      primary  \n",
       "10         1.0  33492891.0  33708876.0   33600883.0           -      primary  \n",
       "11         1.0  34001766.0  34052766.0          NaN           -      primary  \n",
       "12         1.0  33947306.0  34346673.0   34146989.0           +    secondary  \n",
       "13         1.0  34000906.0  34051906.0          NaN           +    secondary  \n",
       "14         1.0  34000906.0  34051906.0          NaN           +    secondary  \n",
       "15         1.0  34000894.0  34051894.0          NaN           +    secondary  \n",
       "16         1.0  33995207.0  34046207.0          NaN           +    secondary  \n",
       "17         1.0  33994963.0  34045963.0          NaN           +    secondary  \n",
       "18         1.0  33947306.0  34346673.0   34146989.0           +      primary  \n",
       "19         1.0  34021181.0  34072181.0          NaN           +    secondary  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_pickle(\"results/demo-regions_all-annotations.p\").head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "final-position",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>gtfanno_uid</th>\n",
       "      <th>center</th>\n",
       "      <th>feat_class</th>\n",
       "      <th>perc_feature</th>\n",
       "      <th>perc_region</th>\n",
       "      <th>distance</th>\n",
       "      <th>has_center</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>appris_principal_score</th>\n",
       "      <th>feat_chrom</th>\n",
       "      <th>feat_start</th>\n",
       "      <th>feat_end</th>\n",
       "      <th>feat_center</th>\n",
       "      <th>feat_strand</th>\n",
       "      <th>feature_rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>7139742</td>\n",
       "      <td>7139990</td>\n",
       "      <td>0</td>\n",
       "      <td>7139866.0</td>\n",
       "      <td>DCRD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-19278.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Pcmtd1</td>\n",
       "      <td>ENSMUSG00000051285.18</td>\n",
       "      <td>ENSMUST00000061280.17</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7109144.0</td>\n",
       "      <td>7160144.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>9548071</td>\n",
       "      <td>9548115</td>\n",
       "      <td>1</td>\n",
       "      <td>9548093.0</td>\n",
       "      <td>DCRD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-37492.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Gm18299</td>\n",
       "      <td>ENSMUSG00000104428.2</td>\n",
       "      <td>ENSMUST00000192026.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9509601.0</td>\n",
       "      <td>9560601.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>10316090</td>\n",
       "      <td>10316863</td>\n",
       "      <td>2</td>\n",
       "      <td>10316476.5</td>\n",
       "      <td>DCRD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-47449.5</td>\n",
       "      <td>True</td>\n",
       "      <td>Gm37569</td>\n",
       "      <td>ENSMUSG00000102556.2</td>\n",
       "      <td>ENSMUST00000192111.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10268027.0</td>\n",
       "      <td>10319027.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>10316090</td>\n",
       "      <td>10316863</td>\n",
       "      <td>2</td>\n",
       "      <td>10316476.5</td>\n",
       "      <td>DCRD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-13581.5</td>\n",
       "      <td>True</td>\n",
       "      <td>Arfgef1</td>\n",
       "      <td>ENSMUSG00000067851.12</td>\n",
       "      <td>ENSMUST00000088615.11</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10301895.0</td>\n",
       "      <td>10352895.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>16297792</td>\n",
       "      <td>16297835</td>\n",
       "      <td>3</td>\n",
       "      <td>16297813.5</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-916.5</td>\n",
       "      <td>True</td>\n",
       "      <td>Gm28095</td>\n",
       "      <td>ENSMUSG00000100868.2</td>\n",
       "      <td>ENSMUST00000186343.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>16248730.0</td>\n",
       "      <td>16299730.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>primary</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome     Start       End  gtfanno_uid      center feat_class  \\\n",
       "0          1   7139742   7139990            0   7139866.0       DCRD   \n",
       "1          1   9548071   9548115            1   9548093.0       DCRD   \n",
       "2          1  10316090  10316863            2  10316476.5       DCRD   \n",
       "3          1  10316090  10316863            2  10316476.5       DCRD   \n",
       "5          1  16297792  16297835            3  16297813.5   Promoter   \n",
       "\n",
       "   perc_feature  perc_region  distance has_center gene_name  \\\n",
       "0           NaN          NaN  -19278.0       True    Pcmtd1   \n",
       "1           NaN          NaN  -37492.0       True   Gm18299   \n",
       "2           NaN          NaN  -47449.5       True   Gm37569   \n",
       "3           NaN          NaN  -13581.5       True   Arfgef1   \n",
       "5           NaN          NaN    -916.5       True   Gm28095   \n",
       "\n",
       "                 gene_id          transcript_id  appris_principal_score  \\\n",
       "0  ENSMUSG00000051285.18  ENSMUST00000061280.17                     1.0   \n",
       "1   ENSMUSG00000104428.2   ENSMUST00000192026.2                     0.0   \n",
       "2   ENSMUSG00000102556.2   ENSMUST00000192111.2                     0.0   \n",
       "3  ENSMUSG00000067851.12  ENSMUST00000088615.11                     1.0   \n",
       "5   ENSMUSG00000100868.2   ENSMUST00000186343.2                     0.0   \n",
       "\n",
       "   feat_chrom  feat_start    feat_end  feat_center feat_strand feature_rank  \n",
       "0         1.0   7109144.0   7160144.0          NaN           +      primary  \n",
       "1         1.0   9509601.0   9560601.0          NaN           -      primary  \n",
       "2         1.0  10268027.0  10319027.0          NaN           -      primary  \n",
       "3         1.0  10301895.0  10352895.0          NaN           -      primary  \n",
       "5         1.0  16248730.0  16299730.0          NaN           +      primary  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "primary_annotations = pd.read_pickle(\"results/demo-regions_primary-annotations.p\")\n",
    "primary_annotations.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cubic-scenario",
   "metadata": {},
   "source": [
    "## Merging primary annotations into one annotation per region"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "composed-belief",
   "metadata": {},
   "source": [
    "The primary annotations can contain multiple annotations per region, if multiple genes are annotated to the region at the same precedence level (e.g. if a region is annotated to the promoters of two different genes). In this case, you can either\n",
    "- choose a single annotation, according to one of the available statistics (e.g. the distance)\n",
    "- merge multiple annotations to get one row per gene\n",
    "\n",
    "Picking one gene per region based on the distance would work like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "mechanical-chile",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>gtfanno_uid</th>\n",
       "      <th>center</th>\n",
       "      <th>feat_class</th>\n",
       "      <th>perc_feature</th>\n",
       "      <th>perc_region</th>\n",
       "      <th>distance</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>appris_principal_score</th>\n",
       "      <th>feat_chrom</th>\n",
       "      <th>feat_start</th>\n",
       "      <th>feat_end</th>\n",
       "      <th>feat_center</th>\n",
       "      <th>feat_strand</th>\n",
       "      <th>feature_rank</th>\n",
       "      <th>abs_distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2141</th>\n",
       "      <td>Pcmtd1</td>\n",
       "      <td>1</td>\n",
       "      <td>7139742</td>\n",
       "      <td>7139990</td>\n",
       "      <td>0</td>\n",
       "      <td>7139866.0</td>\n",
       "      <td>DCRD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-19278.0</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSMUSG00000051285.18</td>\n",
       "      <td>ENSMUST00000061280.17</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7109144.0</td>\n",
       "      <td>7160144.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>primary</td>\n",
       "      <td>19278.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1023</th>\n",
       "      <td>Gm18299</td>\n",
       "      <td>1</td>\n",
       "      <td>9548071</td>\n",
       "      <td>9548115</td>\n",
       "      <td>1</td>\n",
       "      <td>9548093.0</td>\n",
       "      <td>DCRD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-37492.0</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSMUSG00000104428.2</td>\n",
       "      <td>ENSMUST00000192026.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9509601.0</td>\n",
       "      <td>9560601.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>primary</td>\n",
       "      <td>37492.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270</th>\n",
       "      <td>Arfgef1</td>\n",
       "      <td>1</td>\n",
       "      <td>10316090</td>\n",
       "      <td>10316863</td>\n",
       "      <td>2</td>\n",
       "      <td>10316476.5</td>\n",
       "      <td>DCRD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-13581.5</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSMUSG00000067851.12</td>\n",
       "      <td>ENSMUST00000088615.11</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10301895.0</td>\n",
       "      <td>10352895.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>primary</td>\n",
       "      <td>13581.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1267</th>\n",
       "      <td>Gm37569</td>\n",
       "      <td>1</td>\n",
       "      <td>10316090</td>\n",
       "      <td>10316863</td>\n",
       "      <td>2</td>\n",
       "      <td>10316476.5</td>\n",
       "      <td>DCRD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-47449.5</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSMUSG00000102556.2</td>\n",
       "      <td>ENSMUST00000192111.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10268027.0</td>\n",
       "      <td>10319027.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>primary</td>\n",
       "      <td>47449.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1133</th>\n",
       "      <td>Gm28095</td>\n",
       "      <td>1</td>\n",
       "      <td>16297792</td>\n",
       "      <td>16297835</td>\n",
       "      <td>3</td>\n",
       "      <td>16297813.5</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-916.5</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSMUSG00000100868.2</td>\n",
       "      <td>ENSMUST00000186343.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>16248730.0</td>\n",
       "      <td>16299730.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>primary</td>\n",
       "      <td>916.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2530</th>\n",
       "      <td>Snrk</td>\n",
       "      <td>9</td>\n",
       "      <td>121896742</td>\n",
       "      <td>121897004</td>\n",
       "      <td>2994</td>\n",
       "      <td>121896873.0</td>\n",
       "      <td>DCRD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-49496.0</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSMUSG00000038145.18</td>\n",
       "      <td>ENSMUST00000120173.9</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>121896369.0</td>\n",
       "      <td>121947369.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>primary</td>\n",
       "      <td>49496.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>Ano10</td>\n",
       "      <td>9</td>\n",
       "      <td>122074412</td>\n",
       "      <td>122074614</td>\n",
       "      <td>2995</td>\n",
       "      <td>122074512.5</td>\n",
       "      <td>intron</td>\n",
       "      <td>0.001704</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10298.5</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSMUSG00000037949.9</td>\n",
       "      <td>ENSMUST00000042546.4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>122004940.0</td>\n",
       "      <td>122123489.0</td>\n",
       "      <td>122064214.0</td>\n",
       "      <td>-</td>\n",
       "      <td>primary</td>\n",
       "      <td>10298.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>446</th>\n",
       "      <td>Cdcp1</td>\n",
       "      <td>9</td>\n",
       "      <td>123078354</td>\n",
       "      <td>123078533</td>\n",
       "      <td>2997</td>\n",
       "      <td>123078443.5</td>\n",
       "      <td>DCRD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-33340.5</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSMUSG00000035498.11</td>\n",
       "      <td>ENSMUST00000039229.8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>123044103.0</td>\n",
       "      <td>123095103.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>primary</td>\n",
       "      <td>33340.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1779</th>\n",
       "      <td>Lars2</td>\n",
       "      <td>9</td>\n",
       "      <td>123186193</td>\n",
       "      <td>123186646</td>\n",
       "      <td>2998</td>\n",
       "      <td>123186419.5</td>\n",
       "      <td>DCRD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-9572.5</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSMUSG00000035202.9</td>\n",
       "      <td>ENSMUST00000038863.9</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>123145992.0</td>\n",
       "      <td>123196992.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>primary</td>\n",
       "      <td>9572.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2416</th>\n",
       "      <td>Scp2-ps2</td>\n",
       "      <td>9</td>\n",
       "      <td>123186193</td>\n",
       "      <td>123186646</td>\n",
       "      <td>2998</td>\n",
       "      <td>123186419.5</td>\n",
       "      <td>DCRD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-49385.5</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSMUSG00000058492.6</td>\n",
       "      <td>ENSMUST00000216438.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>123136034.0</td>\n",
       "      <td>123187034.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>primary</td>\n",
       "      <td>49385.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2898 rows  21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_name Chromosome      Start        End  gtfanno_uid       center  \\\n",
       "2141    Pcmtd1          1    7139742    7139990            0    7139866.0   \n",
       "1023   Gm18299          1    9548071    9548115            1    9548093.0   \n",
       "270    Arfgef1          1   10316090   10316863            2   10316476.5   \n",
       "1267   Gm37569          1   10316090   10316863            2   10316476.5   \n",
       "1133   Gm28095          1   16297792   16297835            3   16297813.5   \n",
       "...        ...        ...        ...        ...          ...          ...   \n",
       "2530      Snrk          9  121896742  121897004         2994  121896873.0   \n",
       "255      Ano10          9  122074412  122074614         2995  122074512.5   \n",
       "446      Cdcp1          9  123078354  123078533         2997  123078443.5   \n",
       "1779     Lars2          9  123186193  123186646         2998  123186419.5   \n",
       "2416  Scp2-ps2          9  123186193  123186646         2998  123186419.5   \n",
       "\n",
       "     feat_class  perc_feature  perc_region  distance  ...  \\\n",
       "2141       DCRD           NaN          NaN  -19278.0  ...   \n",
       "1023       DCRD           NaN          NaN  -37492.0  ...   \n",
       "270        DCRD           NaN          NaN  -13581.5  ...   \n",
       "1267       DCRD           NaN          NaN  -47449.5  ...   \n",
       "1133   Promoter           NaN          NaN    -916.5  ...   \n",
       "...         ...           ...          ...       ...  ...   \n",
       "2530       DCRD           NaN          NaN  -49496.0  ...   \n",
       "255      intron      0.001704          1.0   10298.5  ...   \n",
       "446        DCRD           NaN          NaN  -33340.5  ...   \n",
       "1779       DCRD           NaN          NaN   -9572.5  ...   \n",
       "2416       DCRD           NaN          NaN  -49385.5  ...   \n",
       "\n",
       "                    gene_id          transcript_id appris_principal_score  \\\n",
       "2141  ENSMUSG00000051285.18  ENSMUST00000061280.17                    1.0   \n",
       "1023   ENSMUSG00000104428.2   ENSMUST00000192026.2                    0.0   \n",
       "270   ENSMUSG00000067851.12  ENSMUST00000088615.11                    1.0   \n",
       "1267   ENSMUSG00000102556.2   ENSMUST00000192111.2                    0.0   \n",
       "1133   ENSMUSG00000100868.2   ENSMUST00000186343.2                    0.0   \n",
       "...                     ...                    ...                    ...   \n",
       "2530  ENSMUSG00000038145.18   ENSMUST00000120173.9                    1.0   \n",
       "255    ENSMUSG00000037949.9   ENSMUST00000042546.4                    1.0   \n",
       "446   ENSMUSG00000035498.11   ENSMUST00000039229.8                    1.0   \n",
       "1779   ENSMUSG00000035202.9   ENSMUST00000038863.9                    1.0   \n",
       "2416   ENSMUSG00000058492.6   ENSMUST00000216438.2                    0.0   \n",
       "\n",
       "      feat_chrom   feat_start     feat_end  feat_center  feat_strand  \\\n",
       "2141         1.0    7109144.0    7160144.0          NaN            +   \n",
       "1023         1.0    9509601.0    9560601.0          NaN            -   \n",
       "270          1.0   10301895.0   10352895.0          NaN            -   \n",
       "1267         1.0   10268027.0   10319027.0          NaN            -   \n",
       "1133         1.0   16248730.0   16299730.0          NaN            +   \n",
       "...          ...          ...          ...          ...          ...   \n",
       "2530         9.0  121896369.0  121947369.0          NaN            +   \n",
       "255          9.0  122004940.0  122123489.0  122064214.0            -   \n",
       "446          9.0  123044103.0  123095103.0          NaN            -   \n",
       "1779         9.0  123145992.0  123196992.0          NaN            +   \n",
       "2416         9.0  123136034.0  123187034.0          NaN            -   \n",
       "\n",
       "     feature_rank abs_distance  \n",
       "2141      primary      19278.0  \n",
       "1023      primary      37492.0  \n",
       "270       primary      13581.5  \n",
       "1267      primary      47449.5  \n",
       "1133      primary        916.5  \n",
       "...           ...          ...  \n",
       "2530      primary      49496.0  \n",
       "255       primary      10298.5  \n",
       "446       primary      33340.5  \n",
       "1779      primary       9572.5  \n",
       "2416      primary      49385.5  \n",
       "\n",
       "[2898 rows x 21 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "one_anno_per_region = (\n",
    "    primary_annotations\n",
    "    .assign(abs_distance=lambda df: df['distance'].abs())\n",
    "    .sort_values([\"gene_name\", \"abs_distance\"], ascending=True)\n",
    "    .groupby(\"gene_name\", as_index=False)\n",
    "    .first()\n",
    "    .sort_values(['Chromosome', 'Start', 'End'])\n",
    ")          \n",
    "one_anno_per_region"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:gtfanno_env]",
   "language": "python",
   "name": "conda-env-gtfanno_env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  },
  "toc-autonumbering": true
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
